ipykernel/kernelbase.py

```diff
@@ -90,16 +90,9 @@ class Kernel(SingletonConfigurable):
 
     processes: dict[str, psutil.Process] = {}
 
-    @observe("eventloop")
-    def _update_eventloop(self, change):
-        """schedule call to eventloop from IOLoop"""
-        loop = ioloop.IOLoop.current()
-        if change.new is not None:
-            loop.add_callback(self.enter_eventloop)
-
```

Comment:
Hi, this `observe` was the way the kernel entered the eventloop when the `%gui` magic is used, and I don't see a replacement for it - this functionality is thus broken in my testing. `enable_gui` only sets this variable and relies on the observer to start the loop: https://github.com/davidbrochart/ipykernel/blob/92f979eda8280a8898d5235c35df62ac3ee15c79/ipykernel/eventloops.py#L397

It's maybe a little concerning that this didn't cause any of the eventloop tests to fail?

Reply:
Yes, at https://github.com/ipython/ipykernel/blob/main/tests/test_eventloop.py - they seem to mostly not use the magic (entering the loop explicitly), or not check whether the loop is running (just if the related variables are set up)

Reply:
I'm wondering if we should take advantage of the switch to AnyIO to do event loop integration differently. Trio has a [guest mode](https://trio.readthedocs.io/en/stable/reference-lowlevel.html#what-is-guest-mode) just for that, so that would be the way to go if ipykernel is configured to run with the Trio backend, but then users wouldn't be able to use asyncio. AnyIO doesn't support a guest mode for the asyncio backend, but [it could](https://github.com/agronholm/anyio/issues/251).

Reply:
I'm actually not sure of the top of my head how to test for this better - "is the GUI eventloop running" is a slightly nebulous question... Perhaps firing timers within the loop could be an approach - it's what the Tk test does, though it calls the eventloop method directly rather than through the kernel currently

Reply:
Do i understand it right that trio's guest mode needs trio to run as the guest within some external event loop? Currently we 'host' the GUI loop within the kernel's loop, would trio's approach require us to start the GUI loop and then add a smaller trio loop within that as as a guest?

Reply:
Yes I think you're right.

Reply:
In case it helps, there's also [`aioguest`](https://github.com/oremanj/aioguest)